src/cli/self_update/shell.rs

```diff
@@ -0,0 +1,213 @@
+//!
+//! Duplicating paths in PATH can cause performance issues when the OS searches
+//! the same place multiple times. Traditionally, Unix configurations have
+//! resolved this by setting up PATHs in the shell's login profile.
+//!
+//! This has its own issues. Login profiles are only intended to run once, but
+//! changing the PATH is common enough that people may run it twice. Desktop
+//! environments often choose to NOT start login shells in GUI terminals. Thus,
+//! a trend has emerged to place PATH updates in other run-commands (rc) files,
+//! leaving Rustup with few assumptions to build on for fulfilling its promise
+//! to set up PATH appropriately.
+//!
+//! Rustup addresses this by:
+//! 1) using a shell script that updates PATH if the path is not in PATH
+//! 2) sourcing this script in any known and appropriate rc file
+
+use std::path::PathBuf;
+
+use error_chain::bail;
+
+use super::*;
+use crate::process;
+
+pub type Shell = Box<dyn UnixShell>;
+
+#[derive(Debug, PartialEq)]
+pub struct ShellScript {
+    content: &'static str,
+    name: &'static str,
+}
+
+impl ShellScript {
+    pub fn write(&self) -> Result<()> {
+        let home = utils::cargo_home()?;
+        let cargo_bin = format!("{}/bin", cargo_home_str()?);
+        let env_name = home.join(self.name);
+        let env_file = self.content.replace("{cargo_bin}", &cargo_bin);
+        utils::write_file(self.name, &env_name, &env_file)?;
+        Ok(())
+    }
+}
+
+// TODO: Update into a bytestring.
+pub fn cargo_home_str() -> Result<Cow<'static, str>> {
+    let path = utils::cargo_home()?;
+
+    let default_cargo_home = utils::home_dir()
+        .unwrap_or_else(|| PathBuf::from("."))
+        .join(".cargo");
+    Ok(if default_cargo_home == path {
+        "$HOME/.cargo".into()
+    } else {
+        match path.to_str() {
+            Some(p) => p.to_owned().into(),
+            None => bail!("Non-Unicode path!"),
+        }
+    })
+}
+
+// TODO: Tcsh (BSD)
+// TODO?: Make a decision on Ion Shell, Power Shell, Nushell
+// Cross-platform non-POSIX shells have not been assessed for integration yet
+fn enumerate_shells() -> Vec<Shell> {
+    vec![Box::new(Posix), Box::new(Bash), Box::new(Zsh)]
+}
+
+pub fn get_available_shells() -> impl Iterator<Item = Shell> {
+    enumerate_shells().into_iter().filter(|sh| sh.does_exist())
+}
+
+pub trait UnixShell {
+    // Detects if a shell "exists". Users have multiple shells, so an "eager"
+    // heuristic should be used, assuming shells exist if any traces do.
+    fn does_exist(&self) -> bool;
+
+    // Gives all rcfiles of a given shell that rustup is concerned with.
+    // Used primarily in checking rcfiles for cleanup.
+    fn rcfiles(&self) -> Vec<PathBuf>;
+
+    // Gives rcs that should be written to.
+    fn update_rcs(&self) -> Vec<PathBuf>;
+
+    // Writes the relevant env file.
+    fn env_script(&self) -> ShellScript {
+        ShellScript {
+            name: "env.sh",
```

Comment:
This is the API difference I think.  `env` rather than `env.sh` would be expected by users.

Reply:
Well yes, you had suggested it be changed for shellcheck convenience (and no, I don't think it's viable to have it written one way on our side and one way on the user side, that just injects more confusion). I can change it back.

Reply:
( since all the files will likely be named env with different extensions it's probably fine anyways. )